frame/staking/src/offchain_election.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helpers for offchain worker election.
+
+use crate::{Call, CompactAssignments, Module, SessionInterface, Trait};
+use codec::Encode;
+use frame_system::offchain::{SubmitUnsignedTransaction};
+use sp_phragmen::{reduce, ExtendedBalance, PhragmenResult, StakedAssignment};
+use sp_std::{prelude::*, cmp::Ordering};
+use sp_runtime::{RuntimeAppPublic, RuntimeDebug};
+
+#[derive(RuntimeDebug)]
+pub(crate) enum OffchainElectionError {
+	/// No signing key has been found on the current node that maps to a validators. This node
+	/// should not run the offchain election code.
+	NoSigningKey,
+	/// Phragmen election returned None. This means less candidate that minimum number of needed
+	/// validators were present. The chain is in trouble and not much that we can do about it.
+	FailedElection,
+}
+
+/// Compares two sets of phragmen scores based on desirability and returns true if `that` is
+/// better `this`.
+///
+/// Evaluation is done in a lexicographic manner.
+pub(crate) fn is_score_better(this: [ExtendedBalance; 3], that: [ExtendedBalance; 3]) -> bool {
+	match that
+		.iter()
+		.enumerate()
+		.map(|(i, e)| e.cmp(&this[i]))
+		.collect::<Vec<Ordering>>()
+		.as_slice()
+	{
+		[Ordering::Greater, _, _] => true,
+		[Ordering::Equal, Ordering::Greater, _] => true,
+		[Ordering::Equal, Ordering::Equal, Ordering::Less] => true,
+		_ => false,
+	}
+}
+
+/// The internal logic of the offchain worker of this module.
+pub(crate) fn compute_offchain_election<T: Trait>() -> Result<(), OffchainElectionError> {
+	let validator_keys = T::SessionInterface::keys::<T::KeyType>();
+	let local_keys = T::KeyType::all();
+
+	if let Some((index, ref pubkey)) = local_keys
+		.into_iter()
+		.find_map(|k|
+			validator_keys
```

Comment:
This is implemented as `find_local_keys` in `SubmitSignedTransaction` more efficiently.

Reply:
For the time being, I prefer not using those at all since then the key type returned by `SubmitSignedTransaction:: find_local_keys()` cannot be used for signing.

Reply:
https://github.com/paritytech/substrate/blob/master/frame/system/src/offchain.rs#L245-L257 Not true?

Reply:
I don't get the comment :D ^^ 

What I meant was: the function that you linked @bkchr is returning a `Vec<(AccountId, Public)>` where `Public` comes from `CreateTransaction::Public`, which is not capable of anything other than being `IdentifyAccount`, namely, signing the payload, as manteioned above. 

If we were to only support signed transactions, this would have been perfectly fine to use these abstractions. Mixing the requirements of signed and unsigned, made it waaay too complicated and I did spend some good amount of time trying to use it but it didn't work out well.

Reply:
Yeah, the bounds there are over complicated, We basically have:
`RuntimeAppPublic <-> RuntimeAppPublic::Generic ?> Public`

1. The first one is app-specific wrapper for specific crypto
2. The second one is specific-crypto
3. The third one is runtime-specific Public key type (which may be an aggregate like `MultiSigner`).

However you should be able to use `Signer<Public, Signature>` trait to sign data passing `Public`. This trait is implemented for `RuntimeAppPublic` and does the necessary conversions inside. However the bounds might be an overkill for this case. I hope to simplifiy that pattern in the future though, so will refactor this code when it lands.

Reply:
We can also alter `find_local_keys` to return `app-crypto` public keys too (will require a clone though).